org.osgi.service.feature/src/org/osgi/service/feature/ID.java

```diff
@@ -34,125 +33,58 @@
-     * @param artifactId The artifact ID.
-     * @param version The version.
-     */
-    public ID(String groupId, String artifactId, String version) {
-        this(groupId, artifactId, version, null, null);
-    }
+	static final String FEATURE_ID_TYPE = "osgifeature";
 
-    /**
-     * Construct an Artifact ID
-     * @param groupId The group ID.
-     * @param artifactId The artifact ID.
-     * @param version The version.
-     * @param type The type identifier.
-     * @param classifier The classifier.
-     */
-    public ID(String groupId, String artifactId, String version, String type, String classifier) {
-        this.groupId = groupId;
-        this.artifactId = artifactId;
-        this.version = version;
-        this.type = type;
-        this.classifier = classifier;
-    }
-
-    /**
-     * Get the group ID.
-     * @return The group ID.
-     */
-    public String getGroupId() {
-        return groupId;
-    }
-
-    /**
-     * Get the artifact ID.
-     * @return The artifact ID.
-     */
-    public String getArtifactId() {
-        return artifactId;
-    }
-
-    /**
-     * Get the version.
-     * @return The version.
-     */
-    public String getVersion() {
-        return version;
-    }
+	/**
+	 * Get the group ID.
+	 * 
+	 * @return The group ID.
+	 */
+	String getGroupId();
 
-    /**
-     * Get the type identifier.
-     * @return The type identifier.
-     */
-    public Optional<String> getType() {
-        return Optional.ofNullable(type);
-    }
+	/**
+	 * Get the artifact ID.
+	 * 
+	 * @return The artifact ID.
+	 */
+	String getArtifactId();
 
-    /**
-     * Get the classifier.
-     * @return The classifier.
-     */
-    public Optional<String> getClassifier() {
-        return Optional.ofNullable(classifier);
-    }
+	/**
+	 * Get the version.
+	 * 
+	 * @return The version.
+	 */
+	String getVersion();
 
-    @Override
-    public int hashCode() {
-        return Objects.hash(artifactId, classifier, groupId, type, version);
-    }
+	/**
+	 * Get the type identifier.
+	 * 
+	 * @return The type identifier.
+	 */
+	Optional<String> getType();
 
-    @Override
-    public boolean equals(Object obj) {
-        if (this == obj)
-            return true;
-        if (!(obj instanceof ID))
-            return false;
-        ID other = (ID) obj;
-        return Objects.equals(artifactId, other.artifactId) && Objects.equals(classifier, other.classifier)
-                && Objects.equals(groupId, other.groupId) && Objects.equals(type, other.type)
-                && Objects.equals(version, other.version);
-    }
+	/**
+	 * Get the classifier.
+	 * 
+	 * @return The classifier.
+	 */
+	Optional<String> getClassifier();
 
-    @Override
-    public String toString() {
-        return groupId + ":" + artifactId + ":" + version;
-    }
+	/**
+	 * This method returns the ID using the following syntax:
+	 * <p>
+	 * {@code groupId ':' artifactId ( ':' type ( ':' classifier )? )? ':' version }
```

Comment:
To be more clear here:

- classifier is included if its value is not `` (empty string).
- type is included if its value is not `jar` or classifier's value is not `` (empty string).